cockatrice/src/SpoilerBackgroundUpdaterThread.cpp

```diff
@@ -0,0 +1,233 @@
+}
+
+void SpoilerBackgroundUpdaterThread::timeoutOccurredTimeToDownloadSpoilers()
+{
+    qDebug() << "Spoiler Timer has finished";
+
+    settingsCache->setDownloadSpoilerLastUpdateTime(QDateTime::currentMSecsSinceEpoch());
+
+    // If this thread is active, then initiate download
+    if (mbIsActiveThread)
+    {
+        qDebug() << "Timer thread is active, so attempting to download spoilers";
+        downloadSpoilersFile(QString(GITHUB_SPOILERS_DOWNLOAD_URL));
+    }
+
+    // Now that we've downloaded the spoiler, reset the timer.
+    // Take it from the top, Jenkins!
+    runTimer(true);
+}
+
+void SpoilerBackgroundUpdaterThread::handleNewTimeInterval()
+{
+    // Cause the timer to stop then restart w/ new interval
+    qDebug() << "Spoiler Timer has new time interval established";
+    runTimer(true);
+}
+
+void SpoilerBackgroundUpdaterThread::runTimer(bool lbStopAndRestart)
+{
+    // Cause a restart of the timer as if we just restart via start() a timeout() will be
+    // thrown which will trigger a download... which we don't want off time!
+    if (lbStopAndRestart)
+    {
+        mpTimerForSpoilers->stop();
+    }
+
+    // Determine how long to wait between updates (Gets the # of hours from settingsCache / convert Mins -> Millis)
+    float lnTimeToWaitBetween = settingsCache->getDownloadSpoilerTimeMinutes() * 60000;
+
+    // How much time is needed until the next update (NOW() - Last Update Time == Time Remaining)
+    long lnLastUpdateTime = settingsCache->getDownloadSpoilerLastUpdateTime();
+    long lnTimeUntilNextUpdate = QDateTime::currentMSecsSinceEpoch() - lnLastUpdateTime;
+    if (lnTimeUntilNextUpdate <= 500 || lnLastUpdateTime == -1)
+    {
+        // The update just occurred less then half a sec ago, so we will restart the timer from the beginning (time zero)
+        mpTimerForSpoilers->start(static_cast<int>(lnTimeToWaitBetween));
+    }
+    else if (lnTimeUntilNextUpdate < lnTimeToWaitBetween)
+    {
+        // If we have checked for an update in the past and it's not yet time to check for another update
+        // Start the clock to count down until lnTimeUntilNextUpdate
+        mpTimerForSpoilers->start(static_cast<int>(lnTimeToWaitBetween - lnTimeUntilNextUpdate));
+    }
+    else
+    {
+        // It's been too long since we last checked (above threshold), so we will start now (time inf)
+        mpTimerForSpoilers->start(0);
+    }
+
+    qDebug() << "Spoiler Timer has started and will finish in" << mpTimerForSpoilers->interval()/60000. << "minutes";
+}
+
+void SpoilerBackgroundUpdaterThread::downloadSpoilersFile(QUrl acUrl)
+{
+    qDebug() << "Starting to download spoiler on NAM";
+    if (mpNetworkAccessManager == nullptr)
+    {
+        mpNetworkAccessManager = new QNetworkAccessManager(this);
+    }
+
+    QNetworkReply *lpNetworkReply = mpNetworkAccessManager->get(QNetworkRequest(acUrl));
+
+    connect(lpNetworkReply, SIGNAL(finished()), this, SLOT(actDownloadFinishedTokensFile()));
```

Comment:
Maybe rename `actDownloadFinishedTokensFile` to `actDownloadFinishedSpoilersFile`